### PR TITLE
Fix pf upgrade and msi building

### PR DESCRIPTION
**Pull request by crazygao**
# Description

No need of extra bat file.
image

`pf upgrade` migrations

Also fix of msi builder.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-devkit/promptflow/scripts/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (wangchao1230) [inline src/promptflow-devkit/promptflow/scripts/__init__.py:2]: wh do we make this public?

Comment (crazygao) [reply src/promptflow-devkit/promptflow/scripts/__init__.py:2]: let me make this folder private.

File: scripts/installer/windows/scripts/generate_dependency.py [hunk 2]
```diff
@@ -34,28 +44,40 @@ def extract_package_names(packages):
...
     dependencies = []
-    install_requires, extras_requires = extract_requirements('../../../../src/promptflow/setup.py')
+    install_requires, extras_requires = extract_requirements(get_git_base_dir() / 'src/promptflow/setup.py')
     install_requires_names = extract_package_names(install_requires)
     dependencies.extend(install_requires_names)
 
     for key in extras_requires:
         extras_require_names = extract_package_names(extras_requires[key])
         dependencies.extend(extras_require_names)
+    dependencies = list(set(dependencies))
```
Comment (YingChen1996) [inline scripts/installer/windows/scripts/generate_dependency.py:74]: It seems we also need to get package in each toml file or maintain a msi seperate toml if pyinstaller support.
I will look into the it later.

File: src/promptflow-devkit/pyproject.toml [hunk 3]
```diff
@@ -101,7 +101,7 @@ requires = ["poetry-core>=1.5.0"]
 build-backend = "poetry.core.masonry.api"
 
 [tool.poetry.scripts]
-pf = "promptflow._cli._pf.entry:main"
+pf = "promptflow._scripts.pf:main"
```
Comment (wangchao1230) [inline src/promptflow-devkit/pyproject.toml:104]: move this to _cli folder

**Review by wangchao1230 — DISMISSED** [review]

**Review by 0mza987 — APPROVED** [review]

File: src/promptflow-devkit/pyproject.toml (hunk 3 above)
Comment (YingChen1996) [reply src/promptflow-devkit/pyproject.toml:104]: promptflow._cli.pf:main ?

File: scripts/installer/windows/scripts/promptflow.spec.jinja2 [hunk 4]
```diff
@@ -2,7 +2,7 @@
 from PyInstaller.utils.hooks import collect_data_files, collect_all, copy_metadata
 
 datas = [('../resources/CLI_LICENSE.rtf', '.'), ('../../../../src/promptflow/NOTICE.txt', '.'),
-('../../../../src/promptflow/promptflow/_sdk/data/executable/', './promptflow/_sdk/data/executable/'),
+('../../../../src/promptflow-devkit/promptflow/_sdk/data/executable/', './promptflow/_sdk/data/executable/'),
```
Comment (YingChen1996) [inline scripts/installer/windows/scripts/promptflow.spec.jinja2:5]: I think maybe we can remove this: ./promptflow/_sdk/data/executable/

File: .github/workflows/build_msi_installer.yml [hunk 5]
```diff
@@ -108,6 +108,19 @@ jobs:
           setupType: promptflow_with_extra
           scriptPath: ${{ env.testWorkingDirectory }}
 
+      - name: Install wheel
```
Comment (YingChen1996) [inline .github/workflows/build_msi_installer.yml:111]: should we delete this: **name: Setup and Install dev promptflow**

and install promptflow-tools in **- name: Install wheel**

File: .github/workflows/build_msi_installer.yml [hunk 6]
```diff
@@ -74,7 +74,7 @@ jobs:
             if ($version -ne $run_version) {
               throw "Version input does not match the version in promptflow package. Version input: $version, version in promptflow package: $run_version"
             }
-          } 
+          }
```
Comment (YingChen1996) [inline .github/workflows/build_msi_installer.yml:77]: I think we  also need update **- name: Install stable promptflow** when all packages are released in pypi

File: scripts/installer/windows/scripts/generate_dependency.py [hunk 7]
```diff
@@ -34,28 +49,41 @@ def extract_package_names(packages):
 def get_package_dependencies(package_name_list):
     dependencies = []
     for package_name in package_name_list:
-        result = subprocess.run(['pip', 'show', package_name], stdout=subprocess.PIPE)
+        if (is_tool('conda')):
+            result = subprocess.run('conda activate root | pip show {}'.format(package_name),
+                                    shell=True, stdout=subprocess.PIPE)
+        else:
+            result = subprocess.run(['pip', 'show', package_name], stdout=subprocess.PIPE)
```
Comment (YingChen1996) [inline scripts/installer/windows/scripts/generate_dependency.py:56]: I think we only can get required package in each promptflow-* now, and can't get packages in [tool.poetry.extras] ?

Comment (YingChen1996) [reply scripts/installer/windows/scripts/generate_dependency.py:56]: maybe can try get toml extra dependecy like：
```
    # get promptflow-** extra packages
    extra_packages = []
    for package in all_packages:
        if package.startswith('promptflow'):
            data = toml.load(get_git_base_dir() / "src" / package / "pyproject.toml")
            extras = data.get("tool", {}).get("poetry", {}).get("extras", {})
            for _, values in extras.items():
                extra_packages.extend(values)
    extra_packages_names = extract_package_names(extra_packages)
    all_packages.extend(extra_packages_names)
```

File: src/promptflow-devkit/promptflow/_cli/pf.py [hunk 8]
```diff
@@ -1,8 +1,17 @@
...
 # ---------------------------------------------------------
+import os
+import sys
+
+
+def main():
+    if os.environ.get("PF_INSTALLER") is None:
+        os.environ["PF_INSTALLER"] = "PIP"
+
+    os.execl(sys.executable, sys.executable, "-m", "promptflow._cli._pf.entry", *sys.argv[1:])
```
Comment (YingChen1996) [inline src/promptflow-devkit/promptflow/_cli/pf.py:13]: Not sure whether it's my local dev environment issue, it seems strange when I run pf command in the branch.

image

File: .github/workflows/build_msi_installer.yml [hunk 9]
```diff
@@ -108,6 +108,19 @@ jobs:
           setupType: promptflow_with_extra
           scriptPath: ${{ env.testWorkingDirectory }}
 
+      - name: Install wheel
+        shell: pwsh
+        run: |
+          Set-PSDebug -Trace 1
+          pip install -r ${{ github.workspace }}/src/promptflow/dev_requirements.txt
+          pip install ${{ github.workspace }}/src/promptflow-tracing
+          pip install ${{ github.workspace }}/src/promptflow-core
```
Comment (YingChen1996) [inline .github/workflows/build_msi_installer.yml:117]: please use promptflow-core[executor-service] here, Not sure why fastapi not installed in step: Setup and Install dev promptflow

Comment (crazygao) [reply .github/workflows/build_msi_installer.yml:117]: fastapi is a testing package?

File: .github/workflows/build_msi_installer.yml [hunk 10]
```diff
@@ -108,6 +108,19 @@ jobs:
           setupType: promptflow_with_extra
           scriptPath: ${{ env.testWorkingDirectory }}
 
+      - name: Setup and Install dev promptflow
+        shell: pwsh
```
Comment (YingChen1996) [inline .github/workflows/build_msi_installer.yml:112]: Please trigger a release pipeline with the latest code and paste the link in the description.

**Review by wangchao1230 — DISMISSED** [review]

**Review by YingChen1996 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**